File: python/examples/icons.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2024 Darby Johnston
# All rights reserved.

import feather_tk as ftk

import sys

# Create the context and application.
context = ftk.Context()
app = ftk.App(context, sys.argv, "icons", "Icons example")
if app.getExit() != 0:
    sys.exit(app.getExit())

# Create a window.
window = ftk.MainWindow(context, app, "icons", ftk.Size2I(1280, 960))

# Create a layout.
layout = ftk.GridLayout(context)
layout.rowBackgroundRole = ftk.ColorRole.Base
layout.spacingRole = ftk.SizeRole.SizeRole_None
scrollWidget = ftk.ScrollWidget(context, ftk.ScrollType.Both)
scrollWidget.border = False
scrollWidget.vStretch = ftk.Stretch.Expanding
scrollWidget.widget = layout
window.setWidget(scrollWidget)

# Create icons.
row = 0
for name in context.getSystemByName("feather_tk::IconSystem").names:
    icon = ftk.Icon(context, name, layout)
    layout.setGridPos(icon, row, 0)
    label = ftk.Label(context, name, layout)
    label.marginRole = ftk.SizeRole.MarginSmall
    layout.setGridPos(label, row, 1)
    row = row + 1

# Show the window and run the application.
window.show()
app.run()

# \bug Need to manually reset the window.
window = None

File: examples/icons/icons.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2024-2025 Darby Johnston
// All rights reserved.

#include <feather-tk/ui/App.h>
#include <feather-tk/ui/GridLayout.h>
#include <feather-tk/ui/Icon.h>
#include <feather-tk/ui/IconSystem.h>
#include <feather-tk/ui/Label.h>
#include <feather-tk/ui/MainWindow.h>
#include <feather-tk/ui/ScrollWidget.h>

#include <feather-tk/core/Format.h>

using namespace feather_tk;

FEATHER_TK_MAIN()
{
    try
    {
        // Create the context and application.
        auto context = Context::create();
        auto app = App::create(context, argc, argv, "icons", "Icons example");
        if (app->getExit() != 0)
            return app->getExit();

        // Create a window.
        auto window = MainWindow::create(
            context,
            app,
            "icons",
            Size2I(1280, 960));

        // Create a layout.
        auto layout = GridLayout::create(context);
        layout->setRowBackgroundRole(ColorRole::Base);
        layout->setSpacingRole(SizeRole::None);
        auto scrollWidget = ScrollWidget::create(context, ScrollType::Both);
        scrollWidget->setBorder(false);
        scrollWidget->setWidget(layout);
        scrollWidget->setVStretch(Stretch::Expanding);
        window->setWidget(scrollWidget);

        // Create icons.
        auto iconSystem = context->getSystem<IconSystem>();
        int row = 0;
        for (const auto& name : iconSystem->getNames())
        {
            auto icon = Icon::create(context, name, layout);
            layout->setGridPos(icon, row, 0);
            auto label = Label::create(context, name, layout);
            label->setMarginRole(SizeRole::MarginSmall);
            layout->setGridPos(label, row, 1);
            ++row;
        }

        // Show the window and run the application.
        window->show();
        app->run();
    }
    catch (const std::exception& e)
    {
        std::cout << "ERROR: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

